cli/commands.go

```diff
@@ -2,44 +2,89 @@ package cli
 
 import (
 	"fmt"
-	"os"
+	"regexp"
 	"strings"
 
 	"github.com/rokka-io/rokka-go/rokka"
 )
 
+type UnknownCommandError string
+
+func (e UnknownCommandError) Error() string {
+	return string(e)
+}
+
 type Command struct {
-	args        []string
-	description string
-	fn          func(*rokka.Client)
+	Args        []string
+	Description string
+	fn          func(*rokka.Client, []string, map[string]string) error
 }
 
 var commands = []Command{
-	Command{[]string{"stackoptions", "list"}, "Show default stack options", func(c *rokka.Client) {
+	Command{[]string{"stackoptions", "list"}, "Show default stack options", func(c *rokka.Client, _ []string, _ map[string]string) error {
 		res, err := c.GetStackoptions()
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "Error getting stack options: %s", err)
-			os.Exit(1)
+			return err
+		}
+		fmt.Println(PrettyJSON(res))
+
+		return nil
+	}},
+	Command{[]string{"organizations", "get", "<name>"}, "Get details of an organization", func(c *rokka.Client, args []string, _ map[string]string) error {
+		res, err := c.GetOrganization(args[0])
+		if err != nil {
+			return err
 		}
-		PrettyPrintJSON(res)
+		fmt.Println(PrettyJSON(res))
+
+		return nil
 	}},
 }
 
 func GetCommands() []Command {
 	return commands
 }
 
-func ExecCommand(c *rokka.Client, args []string) {
-	for _, v := range commands {
-		// TODO: improve this hacky line
-		if strings.Join(v.args, " ") == strings.Join(args, " ") {
-			v.fn(c)
-			return
-		}
-	}
-	fmt.Println("Action not found")
-	fmt.Println("Available actions:")
+func ExecCommand(cl *rokka.Client, userArgs []string) error {
 	for _, c := range commands {
-		fmt.Printf("\t%s - %s\n", strings.Join(c.args, " "), c.description)
+		positionalArgs := []string{}
+
+		if len(userArgs) < len(c.Args) {
+			continue
+		}
+
+		for i, arg := range c.Args {
+			// check whether this is a positional argument ("<arg>")
+			isPositionalArg, err := regexp.MatchString("^<.*>", arg)
+			if err != nil {
+				return err
+			}
+
+			if isPositionalArg {
+				positionalArgs = append(positionalArgs, userArgs[i])
+			} else if arg != userArgs[i] {
+				// user provided argument doesn't match
+				break
+			}
+
+			// we reached the end
+			if len(c.Args) == i+1 {
+				options := map[string]string{}
+
+				// parse rest into options ("key=value")
+				if len(userArgs) >= i {
+					for _, option := range userArgs[i:] {
+						split := strings.Split(option, "=")
+						if len(split) == 2 && split[0] != "" && split[1] != "" {
+							options[split[0]] = split[1]
+						}
+					}
+				}
+
+				return c.fn(cl, positionalArgs, options)
+			}
+		}
 	}
+
+	return UnknownCommandError("Unknown command \"" + strings.Join(userArgs, " ") + "\"")
```

Comment:
maybe use `fmt.Sprintf()` instead of manual concat. You could also use [raw string literal (back quote)](https://golang.org/ref/spec#String_literals) in order to not need to escape the double quotes.

Reply:
done